File: src/Modules/Learning/Learning.Domain/Course.cs
using BuildingBlocks.Domain;

namespace Learning.Domain;

public sealed class Course : AggregateRoot
{
    public CourseId Id { get; private set; } = default!;
    public Title Title { get; private set; } = default!;
    public string? Language { get; private set; }

    private Course() { } // EF

    public static Course Create(CourseId id, Title title, string? language = null) => new()
    {
        Id = id ?? throw new ArgumentNullException(nameof(id)),
        Title = title ?? throw new ArgumentNullException(nameof(title)),
        Language = language
    };
}

File: web/projects/learning/src/lib/domain/course.ts
// Frontend domain model for the Learning context. Deliberately thin — the authoritative
// invariants live server-side. Holds only view-model shape + genuinely client-side display state.
export interface Lesson {
  readonly id: string;
  readonly title: string;
  readonly position: number;
  /** Client-side display state: an unpublished lesson renders as locked. */
  readonly isLocked: boolean;
}

export interface Unit {
  readonly id: string;
  readonly title: string;
  readonly position: number;
  readonly lessons: readonly Lesson[];
}

export interface Course {
  readonly id: string;
  readonly title: string;
  readonly language: string | null;
  readonly units: readonly Unit[];
}
